### PR TITLE
ZeRO-3: adaptive prefetch bucket size

### LINKED ISSUE #8193: [REQUEST] Dynamic/Adaptive Prefetching Window for ZeRO-3
**Problem I have Faced**
I am frustrated by the fact that __prefetch_bucket_sz in ZeRO-3 is a static, user-configured value. In large-scale training, network bandwidth can fluctuate and computation step times vary. A static bucket size creates two distinct bottlenecks:

Compute Starvation: If the bucket is too small, the network cannot fetch parameters fast enough, leading to CPU/GPU blocking when param.wait() is called.
Memory Pressure: If the bucket is too large, it aggressively materializes too many parameters in advance, causing memory spikes, eviction thrashing, or OOM errors.

**Solution I thought:**
I would like to introduce an Adaptive/Dynamic Prefetching Window to the PartitionedParameterCoordinator.

The proposed solution involves:

**Stall Tracking:** Using host-side time.time() around the .wait() loop to measure communication wait time, and tracking compute time as the duration between consecutive module fetches.
**Dynamic Scaling:** Calculating a moving average of the stall_ratio (wait time / compute time).
Adjustment Policy:
**Scale Up**: If stall_ratio > 0.05, we scale __prefetch_bucket_sz by a factor (e.g., 1.2x), bounding it to a maximum of __max_n_available_params * 0.5. This aggressively prefetches further ahead in the trace to hide latency.
**Scale Down**: If stall_ratio is near 0.0, we slowly decay the bucket size down to the baseline size to conserve memory pressure when the network is fast enough.

**Alternatives I saw**
The only current alternative is manually tuning prefetch_bucket_size for every specific hardware configuration and model, which does not account for real-time network fluctuations during training. We also considered tying prefetch scaling to CUDA events directly, but determined host-side timing avoids injecting unnecessary synchronization overhead.

**Additional context** Per the new feature contribution guidelines, I am planning the following experiments to demonstrate effectiveness before submitting the Pull Request:

**Task:** Fine-tuning BERT on SQuAD (NLP) and ResNet on CIFAR (CV) using the DeepSpeedExamples repo.
**Metrics to validate:**
Total throughput improvements (samples/sec) compared to the static baseline.
Peak GPU memory allocation (to ensure the dynamic scaling respects memory limits).
Validation loss over a subset of steps to verify convergence correctness is maintained.
I have already prototyped this logic locally and would appreciate feedback from the core team on this design before opening the Pull Requests!

### PR DESCRIPTION
**Pull request by promptsmith1990**
## Problem

The ZeRO-3 prefetch bucket (`stage3_prefetch_bucket_size`) is set once at startup
and never changes. Getting it right requires trial and error: too small and the GPU
sits idle waiting for allgathers to finish; too large and it wastes memory and can
actually hurt throughput by crowding out live parameters.

Issue #8193 asked for a way to adapt the bucket automatically.

## Solution

This PR adds an optional runtime adaptation loop inside
`PartitionedParameterCoordinator` that adjusts `__prefetch_bucket_sz` based on
observed timing:

- **Fetch-wait EMA**: time spent in the per-module wait loop (blocked on inflight
  allgathers). High values mean we didn't prefetch enough.
- **Inter-step EMA**: wall time between consecutive `_fetch_sub_module_impl` calls.
  Used as a proxy for per-module compute time.

Every 10 steps the ratio `wait_ema / step_ema` is checked:
- `> 0.15` → grow the bucket by 25% (GPU is stalling, need more prefetch)
- `< 0.05` → shrink by 10% (barely any wait, we're over-fetching)

The bucket is clamped to `[min_size, max_size]` to stay within sensible bounds.
Both EMAs use α=0.1, so the controller reacts to trends rather than noise.

## New config keys (all opt-in, backward-compatible defaults)

```json
"zero_optimization": {
  "stage3_adaptive_prefetch_bucket_size": true,
  "stage3_adaptive_prefetch_min_size": 10000000,
  "stage3_adaptive_prefetch_max_size": 500000000
}
```

`stage3_prefetch_bucket_size` continues to work as before and is used as the
starting value when adaptation is enabled.

## Files changed

- `deepspeed/runtime/zero/config.py` — three new `DeepSpeedZeroConfig` fields
- `deepspeed/runtime/zero/parameter_offload.py` — reads config, passes to coordinator
- `deepspeed/runtime/zero/partitioned_param_coordinator.py` — EMA tracking and
  `__update_adaptive_prefetch` method; `import time` added; timing point added
  around the fetch-wait loop

## Notes

- Feature is disabled by default; existing behaviour is completely unchanged
- `time.perf_counter()` is CPU-only and adds no CUDA synchronisation
- The adaptation is skipped until the trace is complete (same guard as prefetch)
- Logging is done at INFO level each time the bucket is resized

Resolves #8193

### CODE REVIEW

**Comment by 0z5a** [comment]
Hi @promptsmith1990, I built a follow-up on this PR's current head (`481c21a7`) and would like to collaborate as a co-author on the adaptive-prefetch work.

The DCO-signed-off follow-up commit is [6e70233](https://github.com/0z5a/DeepSpeed/commit/6e7023303c991802216a80cd827f72324b44319a). It fixes raw engine-config plumbing (including inference), makes the resize decision consistent across data-parallel ranks, samples CUDA compute-stream events without synchronizing each fetch, resets stale trace samples, and enforces the initial/runtime bounds. The feature remains opt-in. It can be cherry-picked onto this branch; I have kept this as a follow-up to your PR.

Validation on that exact source: 43 CPU/Gloo tests passed, both single-GPU overlap configurations passed, and 5 two-GPU NCCL tests passed. The native training cases compare 22 steps and final weights against an unpartitioned SGD reference. Added boundary regressions reproduce 3 failures on the original head.

I also completed a BERT-base/SQuAD 1.1 fixed-subset comparison on 2 × RTX 4000 Ada (20 GiB each, PCIe, no NVLink), Python 3.12.14 / PyTorch 2.13.0+cu130 / Transformers 4.57.6. Each policy has three paired trials (seeds 1337/1338/1339), with identical initial weights, data and RNG; policy order rotates across seeds. BF16, ZeRO-3, overlap enabled, AdamW, global batch 4, sequence length 256, 20 warmup + 80 measured training steps. Training/validation subsets contain 512/64 QA examples.

| Policy | Median samples/s | Max allocated GiB/GPU | Max reserved GiB/GPU |
|---|---:|---:|---:|
| Static 1M | 5.580 | 1.458 | 2.178 |
| Adaptive from 1M | 5.706 | 1.458 | 2.334 |
| Static 50M | 5.803 | 1.458 | 2.318 |

All three policies have identical initial/final validation loss, identical global-mean loss at every training step, and bitwise-identical final BF16 parameters within each seed. Final validation losses are 3.011719 / 2.895264 / 3.077637 for seeds 1337/1338/1339, respectively. Both ranks use the same window throughout each run; the adaptive window grows to about 9.31M by the final training step.

Relative to static 1M, the paired adaptive throughput differences are +1.48%, +4.95%, +2.25% (median +2.25%). Relative to static 50M, they are -0.88%, +0.93%, -1.67% (median -0.88%). There is no allocated-memory reduction versus static 1M, and reserved memory is 160 MiB higher. These measurements support correctness on this workload, not a general speedup claim.

Measurement limits: this is a custom fixed-subset harness, not an execution of the full DeepSpeedExamples training script. Inputs are GPU-resident before timing; throughput includes indexing, training and controller work, excludes preprocessing/evaluation, and uses the slower rank. Memory is the measured PyTorch allocator peak, not every driver/NCCL allocation. Controller overhead is not separately isolated. Initial validation advances the controller cadence before warmup. This is a short-run comparison, not a full convergence result.

Would you be open to incorporating this follow-up and collaborating on the remaining design/validation? If you prefer to squash it into your work, please retain my contribution as `Co-authored-by: 0z5a <dezhen.lu@student.uni-tuebingen.de>`.

**Comment by promptsmith1990** [comment]
> Hi @promptsmith1990, I built a follow-up on this PR's current head (`481c21a7`) and would like to collaborate as a co-author on the adaptive-prefetch work.
> 
> The DCO-signed-off follow-up commit is [6e70233](https://github.com/0z5a/DeepSpeed/commit/6e7023303c991802216a80cd827f72324b44319a). It fixes raw engine-config plumbing (including inference), makes the resize decision consistent across data-parallel ranks, samples CUDA compute-stream events without synchronizing each fetch, resets stale trace samples, and enforces the initial/runtime bounds. The feature remains opt-in. It can be cherry-picked onto this branch; I have kept this as a follow-up to your PR.
> 
> Validation on that exact source: 43 CPU/Gloo tests passed, both single-GPU overlap configurations passed, and 5 two-GPU NCCL tests passed. The native training cases compare 22 steps and final weights against an unpartitioned SGD reference. Added boundary regressions reproduce 3 failures on the original head.
> 
> I also completed a BERT-base/SQuAD 1.1 fixed-subset comparison on 2 × RTX 4000 Ada (20 GiB each, PCIe, no NVLink), Python 3.12.14 / PyTorch 2.13.0+cu130 / Transformers 4.57.6. Each policy has three paired trials (seeds 1337/1338/1339), with identical initial weights, data and RNG; policy order rotates across seeds. BF16, ZeRO-3, overlap enabled, AdamW, global batch 4, sequence length 256, 20 warmup + 80 measured training steps. Training/validation subsets contain 512/64 QA examples.
> 
> Policy	Median samples/s	Max allocated GiB/GPU	Max reserved GiB/GPU
> Static 1M	5.580	1.458	2.178
> Adaptive from 1M	5.706	1.458	2.334
> Static 50M	5.803	1.458	2.318
> All three policies have identical initial/final validation loss, identical global-mean loss at every training step, and bitwise-identical final BF16 parameters within each seed. Final validation losses are 3.011719 / 2.895264 / 3.077637 for seeds 1337/1338/1339, respectively. Both ranks use the same window throughout each run; the adaptive window grows to about 9.31M by the final training step.
> 
> Relative to static 1M, the paired adaptive throughput differences are +1.48%, +4.95%, +2.25% (median +2.25%). Relative to static 50M, they are -0.88%, +0.93%, -1.67% (median -0.88%). There is no allocated-memory reduction versus static 1M, and reserved memory is 160 MiB higher. These measurements support correctness on this workload, not a general speedup claim.
> 
> Measurement limits: this is a custom fixed-subset harness, not an execution of the full DeepSpeedExamples training script. Inputs are GPU-resident before timing; throughput includes indexing, training and controller work, excludes preprocessing/evaluation, and uses the slower rank. Memory is the measured PyTorch allocator peak, not every driver/NCCL allocation. Controller overhead is not separately isolated. Initial validation advances the controller cadence before warmup. This is a short-run comparison, not a full convergence result.
> 
> Would you be open to incorporating this follow-up and collaborating on the remaining design/validation? If you prefer to squash it into your work, please retain my contribution as `Co-authored-by: 0z5a <dezhen.lu@student.uni-tuebingen.de>`.

hi! absolutely!

**Comment by 0z5a** [comment]
Thanks, @promptsmith1990! Could you cherry-pick [6e70233](https://github.com/0z5a/DeepSpeed/commit/6e7023303c991802216a80cd827f72324b44319a) into this PR's branch? It is a single DCO-signed-off follow-up whose parent is still the current PR head, `481c21a7`. I tried pushing it directly, but GitHub rejected the push because my account does not have write access to your fork.

From your DeepSpeed checkout:

```bash
git switch feat/zero3-adaptive-prefetch-bucket-size
git fetch https://github.com/0z5a/DeepSpeed.git codex/zero3-adaptive-prefetch-8193
git cherry-pick 6e7023303c991802216a80cd827f72324b44319a
git push origin feat/zero3-adaptive-prefetch-bucket-size
```

The commit fixes engine-config plumbing for training and inference, shares resize decisions across data-parallel ranks, measures compute-stream stalls with device events, discards stale trace samples, and enforces initial/runtime bounds. It includes tests and configuration documentation. I reran the changed-file pre-commit checks; all passed.

The existing validation for this exact commit includes **43 passing CPU/Gloo tests**, both single-GPU overlap configurations, and **5 passing two-GPU NCCL tests**. Native training tests compare 22 steps and final weights against an unpartitioned SGD reference. Three added boundary regressions fail on the original PR head.

For reference, here are the BERT results from the earlier comparison: **BERT-base / SQuAD 1.1**, 2 × RTX 4000 Ada (20 GiB each, PCIe, no NVLink), BF16, ZeRO-3, overlap enabled, AdamW, global batch 4, sequence length 256. Each policy has three paired trials (seeds 1337/1338/1339), with identical initial weights, data and RNG and a rotating policy order. Each run uses 512 training / 64 validation examples, 20 warmup + 80 measured training steps. Environment: Python 3.12.14, PyTorch 2.13.0+cu130, Transformers 4.57.6.

| Policy | Median samples/s | Max allocated GiB/GPU | Max reserved GiB/GPU |
|---|---:|---:|---:|
| Static 1M | 5.580 | 1.458 | 2.178 |
| Adaptive from 1M | 5.706 | 1.458 | 2.334 |
| Static 50M | 5.803 | 1.458 | 2.318 |

Within each seed, all three policies have identical initial/final validation losses, identical global-mean loss at every training step, and bitwise-identical final BF16 parameters. Final validation losses are 3.011719 / 2.895264 / 3.077637 for seeds 1337/1338/1339. Both ranks maintain the same window throughout; the adaptive window reaches about 9.31M by the final training step.

Adaptive throughput changes versus static 1M are +1.48%, +4.95%, and +2.25% across the three seeds (median +2.25%); versus static 50M they are -0.88%, +0.93%, and -1.67% (median -0.88%). Allocated memory is unchanged versus static 1M, while reserved memory is 160 MiB higher.

These are short runs using a custom fixed-subset harness, not full convergence measurements or a general speedup claim. Inputs are GPU-resident before timing; throughput includes batch indexing, training and controller work, excludes preprocessing/evaluation, and uses the slower rank. Memory reports PyTorch allocator peaks, excluding some driver/NCCL allocations. Controller overhead is not separately isolated, and initial validation advances the controller cadence before warmup.
